Source/WebCore/page/writing-tools/WritingToolsController.mm

```diff
@@ -382,24 +382,46 @@
 
     // The character count delta is `sessionRangeCharacterCount - contextTextCharacterCount`;
     // the above check ensures that the full range length expression will never underflow.
-
     auto characterCountDelta = sessionRangeCharacterCount - contextTextCharacterCount;
     auto adjustedCharacterRange = CharacterRange { range.location, range.length + characterCountDelta };
     auto resolvedRange = resolveCharacterRange(sessionRange, adjustedCharacterRange);
 
-    m_page->chrome().client().addSourceTextAnimation(session.identifier, range);
-
     // Prefer using any attributes that `attributedText` may have; however, if it has none,
     // just conduct the replacement so that it matches the style of its surrounding text.
     //
     // This will always be the case for Smart Replies, which creates it's own attributed text
     // without WebKit providing the attributes.
 
+    RefPtr fragment = createFragment(*document->frame(), attributedText.nsAttributedString().get(), { FragmentCreationOptions::NoInterchangeNewlines, FragmentCreationOptions::SanitizeMarkup });
+    if (!fragment) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
+
     auto commandState = finished ? WritingToolsCompositionCommand::State::Complete : WritingToolsCompositionCommand::State::InProgress;
 
-    replaceContentsOfRangeInSession(*state, resolvedRange, attributedText, commandState);
+    m_page->chrome().client().addSourceTextAnimation(session.identifier, range, [weakPage = WeakPtr { *m_page }, weakThis = WeakPtr { *this }, state, resolvedRange, movedFragement = WTFMove(fragment), attributedText, commandState, identifier = session.identifier, sessionRange]() mutable {
```

Comment:
`WritingToolsController` already has a weak reference to page itself, so you don't need `m_page` in the capture list.